feat: add comprehensive type annotations throughout codebase

timhannifan opened the pull request:
- Add type annotations to all functions, methods, and class attributes
- Use Python 3.11+ syntax (X | None instead of Optional[X])
- Proper typing for NumPy arrays, Matplotlib objects, and PyMC models
- Fix matplotlib axes array handling and return type issues
- Resolve circular imports using TYPE_CHECKING and forward references
- Fix mypy errors and add proper type ignore comments
- Add # noqa: ANN401 for necessary Any usage in **kwargs
- Fix PD013/PD011 linting errors for xarray calls
- Remove pylint comments and ensure linting compliance
- All tests pass and mypy shows zero errors

BREAKING CHANGE: Type annotations added to all public APIs

timhannifan committed:
feat: add comprehensive type annotations throughout codebase

- Add type annotations to all functions, methods, and class attributes
- Use Python 3.11+ syntax (X | None instead of Optional[X])
- Proper typing for NumPy arrays, Matplotlib objects, and PyMC models
- Fix matplotlib axes array handling and return type issues
- Resolve circular imports using TYPE_CHECKING and forward references
- Fix mypy errors and add proper type ignore comments
- Add # noqa: ANN401 for necessary Any usage in **kwargs
- Fix PD013/PD011 linting errors for xarray calls
- Remove pylint comments and ensure linting compliance
- All tests pass, mypy shows zero errors, and pre-commit passes

BREAKING CHANGE: Type annotations added to all public APIs

timhannifan committed:
fix: resolve ruff linting errors in test directory

- Add # noqa comments for test-specific warnings (ANN201, ANN001, D103, F403, F405, S311, PLR2004)
- Add missing example_two_by_two_data fixture to test_goodmans_er.py
- All tests pass and ruff check test/ shows zero errors

timhannifan commented:
Types were intentionally left out of the test function signatures in order to change as little as possible in those files. Let me know if you want those added!

peterrrock2 commented on pyei/__init__.py:
```diff
@@ -1,8 +1,28 @@
-from .r_by_c import *
-from .data import *
+
+# Import main classes and functions
```
```suggestion
```

peterrrock2 commented on MIGRATION_NOTES.md:
This is a comment for later just in case I or another maintainer needs it. The UV migration was accidentally deleted in the last PR, so it is added back in here.

peterrrock2 commented on pyei/greiner_quinn_gibbs_sampling.py:
We can add an S311 ignore to the ruff settings

timhannifan replied:
Ignored and removed references from code.

peterrrock2 commented on pyei/plot_utils.py:
```diff
@@ -9,18 +15,25 @@
 from matplotlib.patches import Rectangle
+from matplotlib.transforms import Transform
+
+# from pyei.r_by_c import RowByColumnEI  # Avoid circular import
```
```suggestion
```

peterrrock2 commented on pyei/r_by_c_utils.py:
```diff
@@ -30,26 +30,29 @@ def check_dimensions_of_input(
             r = group_fractions.shape[0]. If not, plotting labels may be inaccurate.
-            """
+            """,
+                stacklevel=2,
```
Thank you for grabbing these

timhannifan committed:
feat: add py.typed marker for PEP 561 compliance

- Add py.typed file to mark package as type-complete
- Configure hatchling to include py.typed in wheel distributions
- Enables type checkers to recognize inline type annotations when importing pyei

peterrrock2 commented on pyei/goodmans_er.py:
```diff
@@ -156,17 +165,26 @@ def weighted_reg_func(_x, _y, _w):
+                #     seed=None,
+                # ).T
+                # yhat_boots = grid.dot(beta_boots).T  # Unused variable
+                return yhat
```
I think that this breaks the plotting function. The previous version used the yhat_boots (cf. line 189), so if we are just doing type updates in this and not API / output updates, we should just return the `yhat_boots` and modify where `fit_fast` is called. If you have a good reason for changing this over, please just let me know!

timhannifan replied:
Added back, big mistake on my part

peterrrock2 commented on pyei/goodmans_er.py:
Thanks for adding errors in sensible places in here

peterrrock2 commented on pyei/r_by_c.py:
```diff
@@ -608,20 +724,39 @@ def polarization_report(
         else:
-            threshold, percentile, _, groups, candidate = self._calculate_polarization(
+            if threshold is None:
+                raise ValueError("threshold must be provided when percentile is None")
```
I believe that this code isn't reachable. The variable `return_interval` is True when `threshold` is None. We are in the 'else' case here so `threshold` is definitely not `None` at this point.

peterrrock2 replied:
We should probably move the check for exactly one of `percentile` or `threshold` outside of this if-else section (c.f. my comment on the margin report)

timhannifan replied:
Resolved.

peterrrock2 commented on pyei/r_by_c.py:
```diff
@@ -546,18 +651,23 @@ def _calculate_polarization(
+        threshold: float | None = None,
+        percentile: float | None = None,
+        verbose: bool = True,
+    ) -> float | str:
```
```suggestion
    ) -> tuple[float, float] | float:
```

peterrrock2 commented on pyei/r_by_c.py:
```diff
@@ -546,18 +651,23 @@ def _calculate_polarization(
-    ):
+        self,
+        groups: list[int],
+        candidate: int,
```
```suggestion
        candidate: str,
```

Right? That is what the docstring says

peterrrock2 commented on pyei/r_by_c.py:
Unless, I am missing something, I think that there has been some int - string type confusion across the file. Fields like `candidate`, `group`, and `candidates` are listed as ints (or lists if ints) when they should have strings

peterrrock2 replied:
It looks like there is a repeated logic error in here:

```python
        return_interval = threshold is None
        ...

        if return_interval:
            ...
        else:
            if threshold is None:
                raise ValueError("threshold must be provided when percentile is None")
```

It's probably best to change this to check that exactly one of 'threshold' or 'percentile' is checked early. I have an example of what I am thinking in the `margin_report` function.

peterrrock2 commented on pyei/r_by_c.py:
```diff
@@ -431,16 +512,21 @@ def _calculate_margin(self, group, candidates, threshold=None, percentile=None):
             to calculate the associated threshold.
             """
             )
-        return threshold, percentile, samples, group, candidates
+        return threshold
 
     def margin_report(
-        self, group, candidates, threshold=None, percentile=None, verbose=True
-    ):
-        """For a given threshold, return the probability that the margin between
+        self,
+        group: int,
+        candidates: list[int],
+        threshold: float | None = None,
+        percentile: float | None = None,
+        verbose: bool = True,
+    ) -> str:
+        """Calculate margin probability or confidence interval.
+
+        For a given threshold, return the probability that the margin between
         the two candidates preferences in the given demographic group is greater than
-        the threshold
-        OR
-        For a given confidence level, calculate the associated confidence interval
+        the threshold OR for a given confidence level, calculate the associated confidence interval
```
Example rework so that you don't have the unreachable code error:

```python

    def margin_report(
        self,
        group: str,
        candidates: list[str],
        threshold: float | None = None,
        percentile: float | None = None,
        verbose: bool = True,
    ) -> str:
        """Calculate margin probability or confidence interval.

        For a given threshold, return the probability that the margin between
        the two candidates preferences in the given demographic group is greater than
        the threshold OR for a given confidence level, calculate the associated confidence interval
        of the difference between the two candidates preference among the group.
        Exactly one of {percentile, threshold} must be None.

        Parameters:
        -----------
        group: str
            Demographic group in question
        candidates: list of str
            Length 2 vector of candidates upon which to calculate the margin
        threshold: float (optional)
            A specified level for the margin between the two candidates. If specified,
            use the threshold to calculate the percentile (% of samples with a larger margin)
        percentile: float (opetional)
            Between 0 and 100. Used to calculate the equal-tailed interval for the margin between
            the two candidates.
        verbose: bool
            If true, print a report putting margin in context
        """

        if self.candidate_names is None or self.demographic_group_names is None:
            raise ValueError(
                "candidate_names and demographic_group_names must be set before calling margin_report"
            )
        if not all(candidate in self.candidate_names for candidate in candidates):
            raise ValueError(
                f"""candidate names must be in the list of candidate_names provided to fit():
                {self.candidate_names}"""
            )
        if group not in self.demographic_group_names:
            raise ValueError(
                f"""group name must be in the list of demographic_group_names
                provided to fit():
                {self.demographic_group_names}"""
            )

        if not ((threshold is None) ^ (percentile is None)):
            raise ValueError(
                "When generating marigin report, exactly one of 'threshold' and 'percentile' "
                "may be None. "
            )

        return_string = ""
        if threshold is not None:
            threshold = self._calculate_margin(group, candidates, threshold, percentile)
            if verbose:
                print(
                    f"There is a {percentile:.1f}% probability that the difference between"
                    + f" {group}s' preferences for {candidates[0]} and {candidates[1]}"
                    + f" is more than {threshold:.2f}."
                )
            return_string = str(threshold)

        if percentile is not None:
            lower_percentile = (100 - percentile) / 2
            upper_percentile = lower_percentile + percentile
            lower_threshold = self._calculate_margin(
                group, candidates, threshold, upper_percentile
            )
            upper_threshold = self._calculate_margin(
                group, candidates, threshold, lower_percentile
            )

            if verbose:
                print(
                    f"There is a {percentile}% probability that the difference between"
                    + f" {group}s' preferences for {candidates[0]} and {candidates[1]} is"
                    + f" between [{lower_threshold:.2f}, {upper_threshold:.2f}]."
                )
            return_string = f"({lower_threshold:.2f}, {upper_threshold:.2f})"

        return return_string
```

timhannifan replied:
Resolved using your suggestion. Fixed an additional bug in this chunk:
```
        if threshold is not None:
            threshold = self._calculate_margin(group, candidates, threshold, percentile)
            if verbose:
                print(
                    f"There is a {percentile:.1f}% probability that the difference between"
                    + f" {group}s' preferences for {candidates[0]} and {candidates[1]}"
                    + f" is more than {threshold:.2f}."
                )
            return_string = str(threshold)
```

Following the XOR, if threshold is not None, percentile is None.

Replaced with:
```
        if threshold is not None:
            threshold = self._calculate_margin(group, candidates, threshold, percentile=None)
            
            if verbose:
                samples = self._get_margin_samples(group, candidates)
                calculated_percentile = 100 * (samples > threshold).sum() / len(samples)
                print(
                    f"There is a {calculated_percentile:.1f}% probability that the difference between"
                    + f" {group}s' preferences for {candidates[0]} and {candidates[1]}"
                    + f" is more than {threshold:.2f}."
                )
            return_string = str(threshold)
```

peterrrock2 commented on pyei/r_by_c.py:
```diff
@@ -524,10 +621,18 @@ def _calculate_polarization(
+            )
+        candidate_index = self.candidate_names.index(str(candidate))
+        group_index_0 = self.demographic_group_names.index(str(groups[0]))
+        group_index_1 = self.demographic_group_names.index(str(groups[1]))
```
Typecast to str can be removed when the type-signature is updated

peterrrock2 commented on pyei/r_by_c.py:
```diff
@@ -608,20 +724,39 @@ def polarization_report(
                     + f" is more than {threshold:.2f}."
                 )
-            return percentile
+            return actual_percentile / 100.0
```
It looks like there is a logic error here. When verbose is `False` the actual percentile will never be set so you will end up dividing a None by a float

timhannifan replied:
Resolved

```
        self._validate_polarization_inputs(groups, candidate)

        samples = self._get_polarization_samples(groups, candidate)
        actual_percentile = 100 * (samples > threshold).sum() / len(samples)

        if verbose:
            print(
                f"There is a {actual_percentile:.1f}% probability that the difference between"
                + f" the groups' preferences for {candidate} ({groups[0]} - {groups[1]}) "
                + f" is more than {threshold:.2f}."
            )
        return actual_percentile / 100.0
```

peterrrock2 commented on pyei/r_by_c.py:
```diff
@@ -546,18 +651,23 @@ def _calculate_polarization(
-        return threshold, percentile, samples, groups, candidate
+        return threshold
 
     def polarization_report(
```
We should probably just break `polarization_report` into 2 functions with different return types. That is not necessary for this PR, but it's worth noting while it's top-of-mind

timhannifan replied:
Done. 
1. created `_get_polarization_samples` helper, extracts common logic for getting samples
2. created `_validate_polarization_inputs` helper, extracts shared validation logic
3. created `polarization_interval`: single return type
4. created `polarization_percentile`: single return type
5. Updated `polarization_report` as backward-compatible wrapper

peterrrock2 commented on pyei/r_by_c.py:
```diff
@@ -917,35 +1149,67 @@ def plot_margin_kde(
-                group, candidates, threshold, percentile
-            )
+            if threshold is None:
+                raise ValueError("threshold must be provided when percentile is None")
```
Unreachable

timhannifan replied:
Resolved four instances of this in r_by_c.py, lines 1164, 587, 729, 1255. All four had had the same issue. Changed all instances to validate that one of threshold or percentile must be None.

peterrrock2 commented on pyei/r_by_c.py:
It's probably worth adding the bit of code


```python
        candidate_index_0 = self.candidate_names.index(str(candidates[0]))
        candidate_index_1 = self.candidate_names.index(str(candidates[1]))
        group_index = self.demographic_group_names.index(str(group))
        samples = (
            self.sampled_voting_prefs[:, group_index, candidate_index_0]
            - self.sampled_voting_prefs[:, group_index, candidate_index_1]
        )
```

to a helper function. This file repeats it a lot.

timhannifan replied:
Created a helper function `_get_margin_samples` and replaced both occurrences.

peterrrock2 commented on pyei/r_by_c.py:
```diff
@@ -976,37 +1240,55 @@ def plot_polarization_kde(
-            candidate,
+            percentile or 0.0,
+            [str(g) for g in groups],
+            str(candidate),
```
Update to type signature should help with this

peterrrock2 commented on pyei/two_by_two.py:
```diff
@@ -581,6 +643,8 @@ def polarization_report(
             return (lower_threshold, upper_threshold)
         else:
+            if threshold is None:
+                raise ValueError("threshold must be provided when percentile is None")
```
Unreachable

timhannifan replied:
Resolved by changing the guard clause to raise an error if both percentile and threshold are provided.

peterrrock2 commented on pyei/two_by_two.py:
```diff
@@ -706,16 +787,22 @@ def plot_polarization_kde(
             thresholds = [lower_threshold, upper_threshold]
         else:
+            if threshold is None:
+                raise ValueError("threshold must be provided when percentile is None")
```
Unreachable

timhannifan replied:
Resolved by changing the guard clause to raise an error if both percentile and threshold are provided.

peterrrock2 commented on pyei/greiner_quinn_gibbs_sampling.py:
```diff
@@ -236,7 +263,7 @@ def greiner_quinn_gibbs_sample(  # pylint: disable=too-many-locals
 
         # (b) sample theta using Metropolis-Hastings
-        theta_samp = _sample_theta(
+        theta_samp, omega_samp = _sample_theta(
```
Okay, this is a nice way to avoid FPE when you run the inverse back to turn $\theta\rightarrow \omega$, but then we need to also get rid of lines 275-279.

timhannifan replied:
Removed the redundant lines. `_theta_to_omega` is longer called in the working code, only in the tests. Do you want to keep or remove?

peterrrock2 replied:
Let's go ahead and remove it;  it's a simple function and we still have `_omega_to_theta` so we can recover it later even if we forgot that it existed previously

peterrrock2 commented on pyei/greiner_quinn_gibbs_sampling.py:
```diff
@@ -133,30 +133,52 @@ def pyei_greiner_quinn_sample(  # pylint: disable=too-many-locals
-    )  # changing variable name for vote prefernces to match pyei
-    idata = az.from_dict(samples)
+
+    # Ensure 'b' variable exists for consistency with pyei
```
```suggestion
    # Ensure 'b' variable exists for consistency with arviz
```

peterrrock2 commented on pyei/greiner_quinn_gibbs_sampling.py:
```diff
@@ -133,30 +133,52 @@ def pyei_greiner_quinn_sample(  # pylint: disable=too-many-locals
     # convert to InferenceData object
-    for var_name in samples.keys():  # pylint: disable=consider-iterating-dictionary
-        samples[var_name] = np.expand_dims(
-            samples[var_name], 0
+    # Convert samples to proper format for InferenceData
```
Remove one of the comments before this code block, please

peterrrock2 reviewed (changes requested):
This looking great so far! Just a few things in the rxc file need to be addressed, and then we'll be good. Thank you again for your work on this!

timhannifan committed:
fix: correct type annotations for candidate and group parameters

- Fix candidate/group parameters from int to str across r_by_c.py methods
- Fix fit_fast return type to tuple[np.ndarray, np.ndarray] in goodmans_er.py
- Fix polarization_report return type annotation
- Remove unnecessary str() conversions
- Add S311 to ruff ignore list for PRNG usage
- Clean up commented code and noqa directives

timhannifan committed:
fix: refactor polarization/margin functions and fix unreachable code

- Fix unreachable code in polarization_report and plot_polarization_kde (two_by_two.py, r_by_c.py)
- Remove redundant omega_samp calculation in greiner_quinn_gibbs_sampling.py
- Extract helper functions: _get_margin_samples and _get_polarization_samples
- Split polarization_report into polarization_interval and polarization_percentile with backward-compatible wrapper
- Refactor margin_report to use XOR validation and fix percentile calculation bug
- Fix type errors by using len(samples) instead of len(self.sampled_voting_prefs)

peterrrock2 commented on pyei/r_by_c.py:
```diff
@@ -448,10 +448,44 @@ def calculate_summary(self) -> None:
+        np.ndarray
+            Samples of candidate[0] - candidate[1] margin for the given group
+        """
+        if self.candidate_names is None or self.demographic_group_names is None:
+            raise ValueError(
+                "candidate_names and demographic_group_names must be set before calling _get_margin_samples"
+            )
+        if self.sampled_voting_prefs is None:
+            raise ValueError(
+                "sampled_voting_prefs must be set before calling _get_margin_samples"
+            )
+
+        candidate_index_0 = self.candidate_names.index(str(candidates[0]))
+        candidate_index_1 = self.candidate_names.index(str(candidates[1]))
+        group_index = self.demographic_group_names.index(str(group))
+
+        samples = (
+            self.sampled_voting_prefs[:, group_index, candidate_index_0]
+            - self.sampled_voting_prefs[:, group_index, candidate_index_1]
+        )
+        return samples
```
It's probably worth spitting out a warning in here if `len(candidates) > 2`

peterrrock2 commented on pyei/r_by_c.py:
```diff
@@ -516,8 +541,8 @@ def _calculate_margin(
-        group: int,
-        candidates: list[int],
+        group: str,
+        candidates: list[str],
```
We should check that the length of this is 2 at the top of the function and spit out a warning if not

peterrrock2 commented on pyei/r_by_c.py:
```diff
@@ -564,16 +587,36 @@ def margin_report(
+                group, candidates, threshold, percentile=None
+            )
+
+            if verbose:
+                samples = self._get_margin_samples(group, candidates)
+                calculated_percentile = 100 * (samples > threshold).sum() / len(samples)
+                print(
+                    f"There is a {calculated_percentile:.1f}% probability that the difference between"
+                    + f" {group}s' preferences for {candidates[0]} and {candidates[1]}"
+                    + f" is more than {threshold:.2f}."
+                )
```
Good catch on this

peterrrock2 reviewed (changes requested):
Just a couple of small things! Thanks for your work!
